File: AULA 2/C#/E - Counterfeit Dollar.cs
using System;
using System.Linq;
using System.Collections.Generic;

public static class Program
{
	public static void Main()
	{
		var result = "";
		
		var numCases = int.Parse(Console.ReadLine());
		
		var dollarDict = new Dictionary<char, int>();
		for(var i = 0; i < numCases; i++)
		{
			var dollars = new Dictionary<char, bool>
			{
				{ 'A', false }, { 'B', false }, { 'C', false }, { 'D', false },
				{ 'E', false }, { 'F', false }, { 'G', false }, { 'H', false },
				{ 'I', false }, { 'J', false }, { 'K', false }, { 'L', false }
			};
			var weighings = new List<Weighing>();
			
			for(var j = 0; j < 3; j++)
			{
				var weighingInput = Console.ReadLine().Split(' ');
				
				var leftSide = weighingInput[2] == "down" ? 1 : 0;
				var rightSide = weighingInput[2] == "down" ? 0 : 1;
				var balance = weighingInput[2] == "down" ? "up" : weighingInput[2];
				
				var weighing = new Weighing
				{
					LeftSide = new HashSet<char>(weighingInput[leftSide].ToCharArray()),
					RightSide = new HashSet<char>(weighingInput[rightSide].ToCharArray()),
					Balance = balance
				};
				
				weighings.Add(weighing);
			}
			
			var index = 0;
			foreach(var weighing in weighings)
			{
				if(weighing.Balance == "even")
				{
					foreach(var dollar in weighing.LeftSide) { dollars[dollar] = true; }
					foreach(var dollar in weighing.RightSide) { dollars[dollar] = true; }
				} else {
				    
				    var doubt = Union(weighing.LeftSide, weighing.RightSide);

					foreach(var dollar in dollars) { dollars[dollar.Key] = !doubt.TryGetValue(dollar.Key, out var _) ? true : dollar.Value; }
				    
					var index2 = 0;
					foreach(var weighing2 in weighings)
					{
						if(index == index2 || weighing2.Balance == "even") { index2++; continue; }
						
						foreach(var item in UnionMinusIntersection(weighing.LeftSide, weighing2.LeftSide))
						{
							dollars[item] = true;
						}
						
						foreach(var item in UnionMinusIntersection(weighing.RightSide, weighing2.RightSide))
						{
							dollars[item] = true;
						}
						
						index2++;
					}
				}
				index++;
			}
			
			var falseDollar = dollars.First(d => d.Value == false).Key;
			
			var falseDollarWeigh = "";
			foreach(var weighing in weighings)
			{
				if(weighing.Balance != "up") continue;
				
				falseDollarWeigh = weighing.LeftSide.TryGetValue(falseDollar, out var _) ? "heavy" : "light" ;
			}
			
			result += falseDollar + " is the counterfeit coin and it is " + falseDollarWeigh + ".\n";
		}
		
		Console.Write(result);
	}
	
	public static HashSet<char> UnionMinusIntersection(HashSet<char> set1, HashSet<char> set2)
	{
		var unionMinusIntersection = new HashSet<char>(set1);
        unionMinusIntersection.UnionWith(set2);
        
		var intersection = new HashSet<char>(set1);
        intersection.IntersectWith(set2);

        unionMinusIntersection.ExceptWith(intersection);
		
		return unionMinusIntersection;
	}
	
	public static HashSet<char> Union(HashSet<char> set1, HashSet<char> set2)
	{
		var union = new HashSet<char>(set1);
        union.UnionWith(set2);
		
		return union;
	}
	
	public class Weighing
	{
		public HashSet<char> LeftSide { get; set; }
		public HashSet<char> RightSide { get; set; }
		public string Balance { get; set; }
	}
}

File: AULA 2/Java/E - Counterfeit Dollar.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        StringBuilder result = new StringBuilder();
        
        int numCases = Integer.parseInt(scanner.nextLine());
        
        for (int i = 0; i < numCases; i++) {
            Map<Character, Boolean> dollars = new HashMap<>();
            for (char c = 'A'; c <= 'L'; c++) {
                dollars.put(c, false);
            }

            List<Weighing> weighings = new ArrayList<>();

            for (int j = 0; j < 3; j++) {
                String[] weighingInput = scanner.nextLine().split(" ");
                
                int leftSide = "down".equals(weighingInput[2]) ? 1 : 0;
                int rightSide = "down".equals(weighingInput[2]) ? 0 : 1;
                String balance = "down".equals(weighingInput[2]) ? "up" : weighingInput[2];

                Weighing weighing = new Weighing(
                    new HashSet<>(toCharacterSet(weighingInput[leftSide])),
                    new HashSet<>(toCharacterSet(weighingInput[rightSide])),
                    balance
                );
                
                weighings.add(weighing);
            }

            int index = 0;
            for (Weighing weighing : weighings) {
                if ("even".equals(weighing.balance)) {
                    for (char dollar : weighing.leftSide) {
                        dollars.put(dollar, true);
                    }
                    for (char dollar : weighing.rightSide) {
                        dollars.put(dollar, true);
                    }
                } else {
                    Set<Character> doubt = union(weighing.leftSide, weighing.rightSide);

                    for (Map.Entry<Character, Boolean> dollar : dollars.entrySet()) {
                        if (!doubt.contains(dollar.getKey())) {
                            dollars.put(dollar.getKey(), true);
                        }
                    }

                    int index2 = 0;
                    for (Weighing weighing2 : weighings) {
                        if (index == index2 || "even".equals(weighing2.balance)) {
                            index2++;
                            continue;
                        }

                        for (char item : unionMinusIntersection(weighing.leftSide, weighing2.leftSide)) {
                            dollars.put(item, true);
                        }

                        for (char item : unionMinusIntersection(weighing.rightSide, weighing2.rightSide)) {
                            dollars.put(item, true);
                        }

                        index2++;
                    }
                }
                index++;
            }

            char falseDollar = dollars.entrySet().stream()
                .filter(d -> !d.getValue())
                .map(Map.Entry::getKey)
                .findFirst()
                .orElse(null);

            String falseDollarWeigh = "";
            for (Weighing weighing : weighings) {
                if (!"up".equals(weighing.balance)) continue;

                falseDollarWeigh = weighing.leftSide.contains(falseDollar) ? "heavy" : "light";
            }

            result.append(falseDollar).append(" is the counterfeit coin and it is ").append(falseDollarWeigh).append(".\n");
        }

        System.out.print(result.toString());
        scanner.close();
    }

    public static Set<Character> unionMinusIntersection(Set<Character> set1, Set<Character> set2) {
        Set<Character> unionMinusIntersection = new HashSet<>(set1);
        unionMinusIntersection.addAll(set2);

        Set<Character> intersection = new HashSet<>(set1);
        intersection.retainAll(set2);

        unionMinusIntersection.removeAll(intersection);

        return unionMinusIntersection;
    }

    public static Set<Character> union(Set<Character> set1, Set<Character> set2) {
        Set<Character> union = new HashSet<>(set1);
        union.addAll(set2);

        return union;
    }

    public static Set<Character> toCharacterSet(String s) {
        Set<Character> set = new HashSet<>();
        for (char c : s.toCharArray()) {
            set.add(c);
        }
        return set;
    }

    static class Weighing {
        Set<Character> leftSide;
        Set<Character> rightSide;
        String balance;

        Weighing(Set<Character> leftSide, Set<Character> rightSide, String balance) {
            this.leftSide = leftSide;
            this.rightSide = rightSide;
            this.balance = balance;
        }
    }
}